sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -622,10 +646,24 @@ object StartsWithExpressionBuilder extends StringBinaryPredicateExpressionBuilde
   }
 }
 
-case class StartsWith(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.startsWith(r)
+case class StartsWith(left: Expression, right: Expression) extends StringPredicate
+  with CollationCheck {
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationId = checkCollationIds()
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      l.startsWith(r)
```

Comment:
Let's think about naming here. I don't like the idea that startsWith without collation implies binary check. Can we call these methods `startsWithBinary`/`containsBinary` etc.?

Reply:
You're right. "startsWith" was the original name for the binary method. Now, it would probably be best to change it to "startsWithBinary" (and probably "startsWithNonBinary" for the other one)

Reply:
Going with:
(old / current implementation - binary collations only) `containsBinary(substring)`
(new implementation - can be used for all collations) `contains(substring, collationId)`